frontend/public/extend/devconsole/shared/utils/visualization.ts

```diff
@@ -0,0 +1,125 @@
+/* eslint-disable no-unused-vars, no-undef */
+interface Task {
+  resources?: Resources;
+  runAfter?: string[];
+  name: string;
+  taskRef: { name: string };
+}
+
+interface Resources {
+  inputs: Resource[];
+  outputs: Resource[];
+}
+
+interface Resource {
+  name: string;
+  resource: string;
+  from?: string;
+}
+
+export const conditions = {
+  hasFromDependency: (task: Task): boolean =>
+    task.hasOwnProperty('resources') &&
+    task.resources.hasOwnProperty('inputs') &&
+    task.resources.inputs[0].hasOwnProperty('from'),
+  hasRunAfterDependency: (task: Task): boolean =>
+    task.hasOwnProperty('runAfter') && task.runAfter.length > 0,
+};
+
+//to be used by both Pipeline and Pipelinerun visualisation
+const sortTasksByRunAfterAndFrom = (tasks: Task[]): Task[] => {
+  //check and sort tasks by 'runAfter' and 'from' dependency
+  const output = tasks;
+  for (let i = 0; i < output.length; i++) {
+    if (conditions.hasRunAfterDependency(output[i])) {
+      let flag = -1;
+      for (let j = 0; j < output.length; j++) {
+        if (i < j && output[j].taskRef.name === output[i].runAfter[output[i].runAfter.length - 1]) {
+          flag = j;
+        }
+      }
+      if (flag > -1) {
+        //swap with last matching task
+        const temp = output[flag];
+        output[flag] = output[i];
+        output[i] = temp;
+      }
+    } else if (conditions.hasFromDependency(output[i])) {
+      let flag = -1;
+      for (let j = 0; j < output.length; j++) {
+        if (i < j && output[j].taskRef.name === output[i].resources.inputs[0].from[0]) {
+          flag = j;
+        }
+      }
+      if (flag > -1) {
+        //swap with last matching task
+        const temp = output[flag];
+        output[flag] = output[i];
+        output[i] = temp;
+      }
+    }
+  }
+  return output;
+};
+
+export const getPipelineTasks = (actualTasks: Task[]): Task[][] => {
+  //P.S: Each element in out array is referred to as stage
+  const out = [];
+  //Step 1: Sort Tasks to get in correct order
+  const tasks = sortTasksByRunAfterAndFrom(actualTasks);
+
+  //Step 2: Push all nodes without any dependencies in different stages
+  tasks.map((task) => {
```

Comment:
```suggestion
  tasks.forEach((task) => {
```
Then you can remove your empty `return` statement.

Reply:
done